src/coreclr/vm/interpexec.cpp

```diff
@@ -1493,6 +1548,121 @@ do {                                                                           \
+                }
+                case INTOP_STELEM_VT_NOREF:
+                {
+                    BASEARRAYREF arrayRef = LOCAL_VAR(ip[1], BASEARRAYREF);
+                    if (arrayRef == NULL)
+                        COMPlusThrow(kNullReferenceException);
+
+                    ArrayBase* arr = (ArrayBase*)OBJECTREFToObject(arrayRef);
+                    uint32_t len = arr->GetNumComponents();
+                    uint32_t idx = (uint32_t)LOCAL_VAR(ip[2], int32_t);
+                    if (idx >= len)
+                        COMPlusThrow(kIndexOutOfRangeException);
+
+                    uint8_t* pData = arr->GetDataPtr();
+                    size_t elemSize = ip[4];
+                    void* elemAddr = pData + idx * elemSize;
+
+                    memcpyNoGCRefs(elemAddr, stack + ip[3], elemSize);
+                    ip += 5;
+                    break;
+                }
+                case INTOP_LDELEMA:
+                {
+                    BASEARRAYREF arrayRef = LOCAL_VAR(ip[2], BASEARRAYREF);
+                    if (arrayRef == NULL)
+                        COMPlusThrow(kNullReferenceException);
+
+                    ArrayBase* arr = (ArrayBase*)OBJECTREFToObject(arrayRef);
+                    uint32_t len = arr->GetNumComponents();
+                    uint32_t idx = (uint32_t)LOCAL_VAR(ip[3], int32_t);
+                    if (idx >= len)
+                        COMPlusThrow(kIndexOutOfRangeException);
+
+                    uint8_t* pData = arr->GetDataPtr();
+                    size_t elemSize = ip[4];
+                    void* elemAddr = pData + idx * elemSize;
+                    LOCAL_VAR(ip[1], void*) = elemAddr;
+                    ip += 5;
+                    break;
+                }
+                case INTOP_LDELEMA_REF:
+                {
+                    BASEARRAYREF arrayRef = LOCAL_VAR(ip[2], BASEARRAYREF);
+                    if (arrayRef == NULL)
+                        COMPlusThrow(kNullReferenceException);
+
+                    ArrayBase* arr = (ArrayBase*)OBJECTREFToObject(arrayRef);
+                    uint32_t len = arr->GetNumComponents();
+                    uint32_t idx = (uint32_t)LOCAL_VAR(ip[3], int32_t);
+                    if (idx >= len)
+                        COMPlusThrow(kIndexOutOfRangeException);
+
+                    uint8_t* pData = arr->GetDataPtr();
+                    size_t elemSize = ip[4];
+                    void* elemAddr = pData + idx * elemSize;
+
+                    MethodTable* arrayElemMT = arr->GetArrayElementTypeHandle().AsMethodTable();
+                    MethodTable* expectedMT = (MethodTable*)pMethod->pDataItems[ip[5]];
+                    if (arrayElemMT != expectedMT)
+                    {
+                        OBJECTREF elemRef = *(OBJECTREF*)elemAddr;
+                        if (!ObjIsInstanceOf(OBJECTREFToObject(elemRef), expectedMT))
+                            COMPlusThrow(kArrayTypeMismatchException);
```

Comment:
```suggestion
                        COMPlusThrow(kArrayTypeMismatchException);
```
LDELEMA_REF should just check an exact match.